components/brave_wallet/hd_key.cc

```diff
@@ -0,0 +1,411 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/hd_key.h"
+
+#include "base/check.h"
+#include "base/logging.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_split.h"
+#include "brave/third_party/bitcoin-core/src/src/base58.h"
+#include "brave/third_party/bitcoin-core/src/src/crypto/ripemd160.h"
+#include "crypto/sha2.h"
+#include "third_party/boringssl/src/include/openssl/hmac.h"
+
+namespace brave_wallet {
+
+namespace {
+constexpr char kMasterSecret[] = "Bitcoin seed";
+#define SERIALIZATION_LEN 78
+#define HARDENED_OFFSET 0x80000000
+#define MAINNET_PUBLIC 0x0488B21E
+#define MAINNET_PRIVATE 0x0488ADE4
+}  // namespace
+
+HDKey::HDKey()
+    : depth_(0),
+      fingerprint_(0),
+      parent_fingerprint_(0),
+      index_(0),
+      identifier_(20),
+      private_key_(0),
+      public_key_(33),
+      chain_code_(32),
+      secp256k1_ctx_(secp256k1_context_create(SECP256K1_CONTEXT_SIGN |
+                                              SECP256K1_CONTEXT_VERIFY)) {}
+HDKey::HDKey(uint8_t depth, uint32_t parent_fingerprint, uint32_t index)
+    : depth_(depth),
+      fingerprint_(0),
+      parent_fingerprint_(parent_fingerprint),
+      index_(index),
+      identifier_(20),
+      private_key_(0),
+      public_key_(33),
+      chain_code_(32),
+      secp256k1_ctx_(secp256k1_context_create(SECP256K1_CONTEXT_SIGN |
+                                              SECP256K1_CONTEXT_VERIFY)) {}
+
+HDKey::~HDKey() {
+  secp256k1_context_destroy(secp256k1_ctx_);
+}
+
+// static
+std::unique_ptr<HDKey> HDKey::GenerateFromSeed(
+    const std::vector<uint8_t>& seed) {
```

Comment:
https://en.bitcoin.it/wiki/BIP_0032/ section Master Key Generation states the seed space must be between 128-512 bits. We should be validating the seed is a valid size and fail hard if not valid. Note: according to the spec 256 bits being the "recommended size". Not really sure we care if its 256 or not, but worth noting.

Reply:
fixed